File: backend/PrintForge.Constants/Permissions.cs
namespace PrintForge.Constants;

/// <summary>Bitmask permission flags — must match frontend CPPermissions enum.</summary>
[Flags]
public enum CPPermissions
{
    Delete = 1,
    Update = 2,
    Read = 4,
    Create = 8,
    Hidden = 16
}

public static class PermissionHelper
{
    public const int FullCrud = (int)(CPPermissions.Delete | CPPermissions.Update | CPPermissions.Read | CPPermissions.Create);

    /// <summary>Return true when userBits contains all requiredBits.</summary>
    public static bool HasPermission(int userBits, int requiredBits) => (userBits & requiredBits) == requiredBits;
}

File: frontend/src/common_assets/Constants/permissions.ts
/** Bitmask permission flags — must match backend CPPermissions. */
export const CPPermissions = {
  DELETE: 1,
  UPDATE: 2,
  READ: 4,
  CREATE: 8,
  HIDDEN: 16,
} as const;

export type PermissionOp = (typeof CPPermissions)[keyof typeof CPPermissions];

export const FULL_CRUD =
  CPPermissions.DELETE |
  CPPermissions.UPDATE |
  CPPermissions.READ |
  CPPermissions.CREATE;
